apps/common/lib/lexical/ast/analysis/alias.ex

```diff
@@ -1,19 +1,61 @@
 defmodule Lexical.Ast.Analysis.Alias do
+  alias Lexical.Ast
+  alias Lexical.Document
   alias Lexical.Document.Position
+  alias Lexical.Document.Range
 
-  defstruct [:module, :as, :line]
+  defstruct [:module, :as, :range]
 
   @type t :: %__MODULE__{
           module: [atom],
           as: module(),
-          line: Position.line()
+          range: Range.t()
         }
 
-  def new(module, as, line) when is_list(module) and is_atom(as) and line > 0 do
-    %__MODULE__{module: module, as: as, line: line}
+  def new(%Document{} = document, ast, module, as) when is_list(module) do
+    range = range_for_ast(document, ast, module, as)
+    %__MODULE__{module: module, as: as, range: range}
   end
 
   def to_module(%__MODULE__{} = alias) do
     Module.concat(alias.module)
   end
+
+  @implicit_aliases [:__MODULE__, :"@for", :"@protocol"]
+  defp range_for_ast(document, ast, _alias, as) when as in @implicit_aliases do
+    range_for_implicit_alias(document, ast)
+  end
+
+  defp range_for_ast(document, ast, alias, _as) do
+    if List.last(alias) == alias do
+      range_for_implicit_alias(document, ast)
+    else
+      range_for_explicit_alias(document, ast)
+    end
+  end
+
+  defp range_for_explicit_alias(%Document{} = document, ast) do
+    case Ast.Range.fetch(ast, document) do
+      {:ok, %Range{end: end_pos} = range} ->
+        %Range{range | end: %Position{end_pos | character: end_pos.character}}
```

Comment:
I found this pretty hard to parse, and I think it's actually a no-op? Breaking it down:

```elixir
# this update doesn't change end_pos
end_pos == %Position{end_pos | character: end_pos.character}
# which means the range update also doesn't change anything
range == %Range{range | end: end_pos}
```